File: CountCountries.py
from mpi4py import MPI
import csv
import sys
import time

# read csv data
def read_csv(filename):
    data = []
    with open(filename, 'r') as file:
        csv_reader = csv.DictReader(file)
        for row in csv_reader:
            data.append(row)
    return data

filename = "../../migration.csv"

# create connection to MPI
comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

if rank == 0:
    data = read_csv(filename)
    chunks = [[] for _ in range(size)]
    for i, row in enumerate(data):
        chunks[i % size].append(row)
else:
    chunks = None

# start timer
start_time = time.time()
# Scatter chunks to all processes
local_chunk = comm.scatter(chunks, root=0)
print("Process", rank, "received", len(local_chunk), "rows of data")

# Perform local computation (count occurrences of each country)
local_counts = {}
for row in local_chunk:
    country = row['country_of_residence']
    if country in local_counts:
        local_counts[country] += 1
    else:
        local_counts[country] = 1

# Gather local counts to root process
global_counts = comm.gather(local_counts, root=0)

# Compute total counts in root process
if rank == 0:
    total_counts = {}
    for counts in global_counts:
        for country, count in counts.items():
            if country in total_counts:
                total_counts[country] += count
            else:
                total_counts[country] = count
    # end timer
    end_time = time.time()

    print("Counts of countries:")
    for country, count in total_counts.items():
        print(country, ":", count)

    print("Time taken for counting: ", (end_time - start_time) * 1000, " milliseconds")

File: CountCountries.java
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

import java.util.Arrays;
import java.util.List;

import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.count;

public class CountCountries {
    public static void main(String[] args) {
        // Create a SparkSession
        SparkSession spark = SparkSession
                .builder()
                .getOrCreate();

        // Create a JavaSparkContext
        JavaSparkContext sc = new JavaSparkContext(spark.sparkContext());

        // Define the schema
        String schemaString = "year_month,month_of_release,passenger_type,direction,citizenship,visa,country_of_residence,estimate,standard_error,status";
        String[] fields = schemaString.split(",");

        // Convert the fields array into a list
        List<String> fieldList = Arrays.asList(fields);

        StructField[] structFields = sc.parallelize(fieldList).map(fieldName ->
                DataTypes.createStructField(fieldName, DataTypes.StringType, true)
        ).collect().toArray(new StructField[0]);

        StructType schema = DataTypes.createStructType(structFields);

        // Read CSV
        Dataset<Row> sdfData = spark.read()
                .format("csv")
                .option("header", "true")
                .schema(schema)
                .load("../../migration.csv");

        // Explicitly repartition the DataFrame
        sdfData = sdfData.repartition(3);

        // Start timer
        long startTime = System.currentTimeMillis();

        // Count the occurrences of each country of residence
        Dataset<Row> countryCounts = sdfData.groupBy("country_of_residence").agg(count("*").alias("count"));
        countryCounts.show();

        // End timer
        long endTime = System.currentTimeMillis();
        long elapsedTime = endTime - startTime;
        System.out.println("Time taken for counting: " + elapsedTime + " milliseconds");

        spark.stop();
    }
}
